plugins/inputs/tencentcloudcm/discover.go

```diff
@@ -0,0 +1,213 @@
+	discoverTool.Add("QCE/LB_PRIVATE", &LBPrivate{})
+	discoverTool.Add("QCE/CES", &CES{})
+	discoverTool.Add("QCE/DC", &DC{})
+	return discoverTool
+}
+
+// DiscoverMetrics discovers metrics supported by registered products
+func (d *discoverTool) DiscoverMetrics() {
+	// discover metrics once
+	for namespace, p := range d.registry {
+		if d.DiscoveredMetrics == nil {
+			d.DiscoveredMetrics = map[string][]string{}
+		}
+		d.DiscoveredMetrics[namespace] = p.Metrics()
+	}
+}
+
+func (d *discoverTool) discoverObjects(accounts []*Account, endpoint string) error {
+	discoveredObjects := map[string]discoverObject{}
+	for _, account := range accounts {
+		for _, namespace := range account.Namespaces {
+			for _, region := range namespace.Regions {
+				// skip discover if instances are explicitly specified
+				if len(region.Instances) != 0 {
+					continue
+				}
+				p, ok := d.registry[namespace.Name]
+				if !ok {
+					continue
+				}
+				discoveredObject, err := Discover(account.crs, region.RegionName, endpoint, p)
+				if err != nil {
+					d.Log.Errorf("discover account:%s region:%s endpoint:%s failed, error: %s",
+						account.Name, region.RegionName, endpoint, err)
+					continue
+				}
+				discoveredObjects[newKey(account.Name, namespace.Name, region.RegionName)] = discoveredObject
+			}
+		}
+	}
+	d.rw.Lock()
+	d.DiscoveredObjects = discoveredObjects
+	d.rw.Unlock()
+	return nil
+}
+
+// Discover discovers instances of registered products
+func (d *discoverTool) Discover(accounts []*Account, interval config.Duration, endpoint string) {
+	ticker := time.NewTicker(time.Duration(interval))
+	defer ticker.Stop()
+
+	err := d.discoverObjects(accounts, endpoint)
+	if err != nil {
+		d.Log.Errorf("discovery failed: %v", err)
+	}
+
+	// discover instances periodically
+	for range ticker.C {
+		err = d.discoverObjects(accounts, endpoint)
+		if err != nil {
+			d.Log.Errorf("discovery failed: %v", err)
+		}
+	}
+}
+
+// GetInstance Get discovered instance detail
+func (d *discoverTool) GetInstance(account, namespace, region, key string) map[string]interface{} {
+	d.rw.RLock()
+	defer d.rw.RUnlock()
+	discoverObject, ok := d.DiscoveredObjects[newKey(account, namespace, region)]
+	if !ok {
+		return nil
+	}
+	instance, ok := discoverObject.Instances[key]
+	if !ok {
+		return nil
+	}
+	return instance
+}
+
+// GetMonitorInstances Get discovered monitor instances
+func (d *discoverTool) GetMonitorInstances(account, namespace, region string) []*monitor.Instance {
+	d.rw.RLock()
+	defer d.rw.RUnlock()
+	discoverObject, ok := d.DiscoveredObjects[newKey(account, namespace, region)]
+	if !ok {
+		return nil
+	}
+	return discoverObject.MonitorInstances
+}
+
+// Discover implements Product interface
+func Discover(crs *common.Credential, region, endpoint string, p Product) (discoverObject, error) {
+	discoverObject := discoverObject{
+		Instances: map[string]map[string]interface{}{},
+	}
+
+	offset, limit := int64(0), int64(100)
```

Comment:
As offset is always zero, please remove it in the calculations and define it as a `const` if you really need it. The same holds for `limit`.

Reply:
fixed in commit 114c8bd

Reply:
I cannot see that commit being applied to this PR...

Reply:
It can be found at the bottom : )

![image](https://user-images.githubusercontent.com/39701004/133275998-afd0a8f2-93a2-48f7-84ec-5c8570134054.png)